### PR TITLE
Create an abstraction on top of the KeyboardioHID interface

### PR DESCRIPTION
**Pull request by obra**
Create an abstraction on top of the KeyboardioHID interface so that user code doesn't need to know anything about it.

(This also paves the way for @wez' EventDispatcher)

@algernon - I'm particularly curious to have you look at my naming. Does this "feel" right?

### CODE REVIEW

File: src/key_events.cpp [hunk 1]
```diff
@@ -40,57 +43,137 @@ static bool handleKeyswitchEventDefault(Key mappedKey, byte row, byte col, uint8
...
+    if (mappedKey.raw == Key_NoKey.raw)
+      return;
+  }
+  mappedKey = Layer.eventHandler(mappedKey, row, col, keyState);
+  if (mappedKey.raw == Key_NoKey.raw)
+    return;
+  handleKeyswitchEventDefault(mappedKey, row, col, keyState);
+}
+
+void pressKeyRaw(Key mappedKey) {
```
Comment (algernon) [inline src/key_events.cpp]: This is one place where I'm a bit conflicted about naming. `pressKeyRaw` does not feel natural... `pressRawKey` or `pressKeyCode` might feel better, but I'm not sure about those, either. Mind you, `pressKeyRaw` does not feel wrong, either. Just... doesn't feel perfect, either.

File: src/key_events.cpp [hunk 2]
```diff
@@ -40,57 +43,137 @@ static bool handleKeyswitchEventDefault(Key mappedKey, byte row, byte col, uint8
...
+  SystemControl.press(mappedKey.keyCode);
+}
+
+void releaseSystem(Key mappedKey) {
+  SystemControl.release();
+}
+
+
+/** Mouse events
+ * See above for commentary on connectionMask. */
```
Comment (algernon) [inline src/key_events.cpp]: `connectionMask`? Are those the buttons? (I do not see a comment about this thing added)

File: src/key_events.cpp [hunk 3]
```diff
@@ -40,57 +43,137 @@ static bool handleKeyswitchEventDefault(Key mappedKey, byte row, byte col, uint8
...
+void pressMouseButtons(uint8_t buttons) {
+  Mouse.press(buttons);
+}
+
+void releaseMouseButtons(uint8_t buttons) {
+  Mouse.release(buttons);
+}
+
+/** Absolute mouse (grapahics tablet) events
+ * See above for commentary on connectionMask. */
```
Comment (algernon) [inline src/key_events.cpp]: Same thing here about `connectionMask`.

File: src/key_events.h [hunk 4]
```diff
@@ -43,6 +43,30 @@ extern const Key keymaps[][ROWS][COLS];
...
+// A facade on top of our HID implementation
 void pressKey(Key mappedKey);
 void releaseKey(Key mappedKey);
+void releaseAllKeys();
+void pressKeyRaw(Key mappedKey);
+void releaseKeyRaw(Key mappedKey);
+
+void pressConsumer(Key mappedKey);
+void releaseConsumer(Key mappedKey);
+
```
Comment (algernon) [inline src/key_events.h]: Reading this a second time, could I suggest `pressConsumerKey` instead? Just to make it crystal clear that we are not trying to press the user. :wink:

File: src/key_events.h [hunk 5]
```diff
@@ -43,6 +43,30 @@ extern const Key keymaps[][ROWS][COLS];
...
+void pressSystem(Key mappedKey);
+void releaseSystem(Key mappedKey);
+
```
Comment (algernon) [inline src/key_events.h]: While this sounds okay-ish, `pressSystemKey` conveys the intent of the functions better for me.

Or, if `pressKeyRaw` remains that, then this should likely be `pressKeySystem`, and the two above `pressKeyConsumer` - and that would feel wrong. So I'm starting to think that `pressKeyRaw` should be `pressRawKey`, and these should be `pressSystemKey` and `pressConsumerKey` above.

**Comment by obra** [comment]
I _Think_ I've addressed everything

**Comment by obra** [comment]
Manually merged.

**Pull request closed without merging**